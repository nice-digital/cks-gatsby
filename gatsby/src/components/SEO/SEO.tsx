import React from "react";
import { Helmet } from "react-helmet";
import { useLocation } from "@reach/router";
import { useSiteMetadata } from "../../hooks/useSiteMetadata";

const defaultDescription =
	"Providing primary care practitioners with a readily accessible summary of the current evidence base and practical guidance on best practice";

interface MetaDatum {
	name: string;
	content: string;
}

interface SEOProps {
	title?: string;
	description?: string;
	noIndex?: boolean;
	additionalMetadata?: Array<MetaDatum>;
}

export const SEO: React.FC<SEOProps> = ({
	title,
	description,
	noIndex,
	additionalMetadata,
}: SEOProps) => {
	const { pathname } = useLocation();
	const { siteUrl } = useSiteMetadata();

	return (
		<>
			<Helmet
				title={title}
				titleTemplate={`%s | CKS | NICE`}
				defaultTitle="CKS | NICE"
			>
				<html lang="en-GB" />
				<meta name="description" content={description || defaultDescription} />
				<meta
					property="og:description"
					content={description || defaultDescription}
				/>
				<meta property="og:url" content={siteUrl + pathname} />
				<meta property="og:locale" content="en_GB" />
				<meta property="og:type" content="website" />
				<meta
					property="og:title"
					content={(title ? `${title} | ` : "") + "CKS | NICE"}
				/>
				<meta property="og:image" content={`${siteUrl}/open-graph-image.png`} />
				<meta property="og:image:width" content="1200" />
				<meta property="og:image:height" content="630" />
				<meta name="twitter:card" content="summary" />
				<meta
					name="twitter:image"
					content={`${siteUrl}/twitter-summary-image.png`}
				/>
				<meta name="twitter:image:alt" content="NICE and CKS logos" />
				<meta name="twitter:site" content="@NICEcomms" />
				<meta name="twitter:creator" content="@NICEcomms" />
				<meta name="theme-color" content="#004650" />
				{additionalMetadata?.map((x, i) => (
					<meta key={i} name={x.name} content={x.content} />
				))}
				{noIndex && <meta name="robots" content="noindex" />}
				<link
					rel="search"
					type="application/opensearchdescription+xml"
					href="/opensearch.xml"
					title="CKS"
				/>
				<link rel="icon" href="/favicon.ico" />
				<link rel="canonical" href={siteUrl + pathname} />
			</Helmet>
		</>
	);
};
