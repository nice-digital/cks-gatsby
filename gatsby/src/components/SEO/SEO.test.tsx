import React from "react";
import { render, waitForDomChange } from "@testing-library/react";

import { SEO } from "./SEO";

describe("SEO", () => {
	it("should set HTML language to UK English", async () => {
		render(<SEO title="Custom title" description="Custom description" />);
		await waitForDomChange();
		expect(document.querySelector("html")?.getAttribute("lang")).toEqual(
			"en-GB"
		);
	});

	it("should set default title and description when none provided", async () => {
		render(<SEO />);

		await waitForDomChange();

		expect(document.querySelector("head")).toMatchSnapshot();
	});

	it("should render custom title and description in document head", async () => {
		render(<SEO title="Custom title" description="Custom description" />);
		await waitForDomChange();

		expect(document.title).toEqual("Custom title | CKS | NICE");
		expect(
			document
				.querySelector("meta[property='og:title']")
				?.getAttribute("content")
		).toEqual("Custom title | CKS | NICE");
		expect(
			document
				.querySelector("meta[name='description']")
				?.getAttribute("content")
		).toEqual("Custom description");
		expect(
			document
				.querySelector("meta[property='og:description']")
				?.getAttribute("content")
		).toEqual("Custom description");
	});
});
