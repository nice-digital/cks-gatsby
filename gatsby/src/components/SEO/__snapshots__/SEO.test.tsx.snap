// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO should set default title and description when none provided 1`] = `
<head>
  <title>
    CKS | NICE
  </title>
  <link
    data-react-helmet="true"
    href="/opensearch.xml"
    rel="search"
    title="CKS"
    type="application/opensearchdescription+xml"
  />
  <link
    data-react-helmet="true"
    href="/favicon.ico"
    rel="icon"
  />
  <link
    data-react-helmet="true"
    href="https://cks.nice.org.uk/test/"
    rel="canonical"
  />
  <meta
    content="Providing primary care practitioners with a readily accessible summary of the current evidence base and practical guidance on best practice"
    data-react-helmet="true"
    name="description"
  />
  <meta
    content="Providing primary care practitioners with a readily accessible summary of the current evidence base and practical guidance on best practice"
    data-react-helmet="true"
    property="og:description"
  />
  <meta
    content="https://cks.nice.org.uk/test/"
    data-react-helmet="true"
    property="og:url"
  />
  <meta
    content="en_GB"
    data-react-helmet="true"
    property="og:locale"
  />
  <meta
    content="website"
    data-react-helmet="true"
    property="og:type"
  />
  <meta
    content="CKS | NICE"
    data-react-helmet="true"
    property="og:title"
  />
  <meta
    content="https://cks.nice.org.uk/open-graph-image.png"
    data-react-helmet="true"
    property="og:image"
  />
  <meta
    content="1200"
    data-react-helmet="true"
    property="og:image:width"
  />
  <meta
    content="630"
    data-react-helmet="true"
    property="og:image:height"
  />
  <meta
    content="summary"
    data-react-helmet="true"
    name="twitter:card"
  />
  <meta
    content="https://cks.nice.org.uk/twitter-summary-image.png"
    data-react-helmet="true"
    name="twitter:image"
  />
  <meta
    content="NICE and CKS logos"
    data-react-helmet="true"
    name="twitter:image:alt"
  />
  <meta
    content="@NICEcomms"
    data-react-helmet="true"
    name="twitter:site"
  />
  <meta
    content="@NICEcomms"
    data-react-helmet="true"
    name="twitter:creator"
  />
  <meta
    content="#004650"
    data-react-helmet="true"
    name="theme-color"
  />
</head>
`;
